ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/io/Pubsub.java

```diff
@@ -0,0 +1,82 @@
+package com.mozilla.telemetry.ingestion.io;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.cloud.pubsub.v1.Subscriber;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.function.Function;
+
+public class Pubsub {
+
+  public static class Read {
+
+    private Subscriber subscriber;
+
+    public Read(String subscriptionName, Function<PubsubMessage, CompletableFuture<?>> output) {
+      ProjectSubscriptionName subscription = ProjectSubscriptionName.parse(subscriptionName);
+      subscriber = Subscriber.newBuilder(subscription,
+          (message, consumer) -> CompletableFuture.supplyAsync(() -> message)
+              .thenComposeAsync(output::apply).whenCompleteAsync((result, exception) -> {
+                if (exception == null) {
+                  consumer.ack();
+                } else {
+                  consumer.nack();
```

Comment:
Should we log exceptions, or record a metric about exceptions?

On that note, do you have any thoughts on metrics for this application, or should we save that for future work? I'm most familiar with dropwizard-metrics as a metrics interface for standalone Java applications, which can emit metrics to Datadog or Stackdriver per configuration.

Reply:
i'm only familiar with statsd and datadog's tagged statsd variant. we should definitely log exceptions, we might initially get away with having stackdriver extract metrics from logs.